Add revisioning of post meta, including ‘footnotes’ by default

adamsilverstein opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/20564

Related Gutenberg Ticket: https://github.com/WordPress/gutenberg/pull/52988

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein committed:
Add revisioning of post meta, including ‘footnotes’ by default

adamsilverstein commented:
See https://github.com/WordPress/gutenberg/pull/52686

adamsilverstein committed:
correct typo in `src/wp-includes/default-filters.php`

adamsilverstein committed:
add tests for meta revisioning

adamsilverstein committed:
Try: revisioning in the rest endpoint directly, after meta is saved

adamsilverstein committed:
Add _wp_save_post_revision_meta helper that copies revisioned meta to last stored revision

adamsilverstein committed:
remove direct REST API integration, move save to wp_after_insert_post

adamsilverstein committed:
Clean up revisions save

adamsilverstein committed:
phpcbf

adamsilverstein committed:
Ensure `$revisions[0]->post_id ` is set before using

adamsilverstein committed:
Test work

adamsilverstein committed:
Expected revision count is 6

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

Mamaduka commented:
@adamsilverstein, should we try using `register_meta` for enabling post meta revisions instead of `wp_post_revision_meta_keys` filter?

adamsilverstein commented:
>  @adamsilverstein, should we try using register_meta for enabling post meta revisions instead of wp_post_revision_meta_keys filter?

Sure, good idea - I'll give it a try

adamsilverstein committed:
Add REST API tests for revisioning of post meta.

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
Remove footnote specific revisioning of meta

adamsilverstein committed:
phpcbf

adamsilverstein committed:
Fix test setup

adamsilverstein committed:
Save revisioned meta in `_wp_put_post_revision`

adamsilverstein committed:
Enable setting revisioning in register_meta

adamsilverstein committed:
phpcs

ellatrix commented on tests/phpunit/tests/rest-api/rest-post-meta-fields.php:
```diff
@@ -3112,4 +3112,93 @@ public function error_delete_query( $query ) {
+		$this->assertSame( 'bar', $response->get_data()['meta']['foo'] );
+
+		// Create two more revisions with different meta values for the foo key.
+		$request = new WP_REST_Request( 'POST', sprintf( '/wp/v2/posts/%d/revisions', $post_id ) );
```
It would be good to test creating revisions through the `/wp/v2/posts/%d` endpoint instead of manually creating revisions.

adamsilverstein replied:
Good suggestion, will add

adamsilverstein replied:
added

adamsilverstein committed:
remove unused $revisioned_keys

adamsilverstein committed:
Add testing for enabling revisions in `register_meta`

adamsilverstein committed:
phpcbf

adamsilverstein commented on src/wp-includes/revision.php:
```diff
@@ -356,6 +356,9 @@ function _wp_put_post_revision( $post = null, $autosave = false ) {
 	}
 
+	// Save any revisioned post meta.
+	wp_save_revisioned_meta_fields( $post['post_parent'], $revision_id );
```
@ellatrix saving the revisioned meta here, directly after saving the revision, solves all of the out-of sync/different saving order issues we previously saw.

ellatrix replied:
I'm not sure this will work for the posts rest api endpoint. But we'll see with a test, and also with the GB e2e test.

adamsilverstein replied:
> I'm not sure this will work for the posts rest api endpoint. But we'll see with a test, and also with the GB e2e test.

Pretty sure it will - tests will confirm!

adamsilverstein replied:
This did require some additional adjustment as you suggested, still working on it a bit. We will need distinct handling for REST requests

adamsilverstein replied:
I've updated this so the meta revision saving is hooked into the `wp_revision_saved` action; that way, we can unhoook the action before the first post save in the REST endpoint where meta hasn't been saved yet.

adamsilverstein committed:
Use a separate global for revisioned keys `$wp_revisioned_meta_keys`

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
Meta test cleanup to match new defaults

adamsilverstein committed:
Adjust expected for `test_should_register_persisted_preferences_meta`

adamsilverstein committed:
Refine REST tests for meta revisioning

adamsilverstein committed:
Avoid duplicate keys in `wp_revisioned_meta_keys`

adamsilverstein committed:
remove unused `original_post_id`

adamsilverstein committed:
REST posts endpoint: update revision meta when saving meta

adamsilverstein committed:
test work

adamsilverstein committed:
Hook `wp_save_revisioned_meta_fields` on `_wp_put_post_revision` so it can be unhooked in the REST API handler

adamsilverstein committed:
Test cleanup

adamsilverstein committed:
Improve footnote display, show number and content

adamsilverstein committed:
Save the revisioned meta directly in the REST endpoint

adamsilverstein committed:
spacing

adamsilverstein commented on src/wp-includes/blocks/footnotes.php:
```diff
@@ -110,6 +110,19 @@ function wp_add_footnotes_to_revision( $fields ) {
+	}
+	$footnotes_html = '';
+	$x = 1;
+	foreach( $footnotes as $footnote ) {
```
cleaner display extracts the footnote content and adds numbering (needs double checking that order is correct)

adamsilverstein replied:
cc: @ellatrix

adamsilverstein commented on src/wp-includes/blocks/footnotes.php:
```diff
@@ -110,6 +110,19 @@ function wp_add_footnotes_to_revision( $fields ) {
+	return $footnotes_html;
 }
-add_filter( 'wp_post_revision_field_footnotes', 'wp_get_footnotes_from_revision', 10, 3 );
+add_filter( '_wp_post_revision_field_footnotes', 'wp_get_footnotes_from_revision', 10, 3 );
```
@ellatrix I think this is hooked incorrectly in GB currently? needs the `_` prefix.

ellatrix replied:
Oh! 🤦‍♀️

adamsilverstein replied:
I see you fixed this in GB; also note display approach here for the footnotes on the revision page in `wp_get_footnotes_from_revision`

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

ellatrix commented on src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -673,6 +673,9 @@ public function create_item( $request ) {
 		}
 
+		// Don't save the post meta revision because they aren't stored yet.
+		remove_action( '_wp_put_post_revision', 'wp_save_revisioned_meta_fields', 10, 2 );
```
I'm not sure I like this. The root of the problem is that the order of filters and actions in this REST API request is wrong. `wp_insert_post` and `wp_update_post` fire the `post_updated` action _after_ meta is updated (through `meta_input`). The REST API should do the same. Why does the REST API only fix the `wp_after_insert_post` to fire after meta is updated? It should do the same for `post_updated` and all those other filters.

ellatrix replied:
I don't know who else has recently worked on these endpoints. Maybe @TimothyBJacobs, @kadamwhite?

ellatrix replied:
I'm also not sure why the REST API decides not to use `meta_input`. Probably because it has limitation, but then those limitations should be addressed? Ideally everything goes through one code path. Since the REST API updates meta by itself, it had to move the `wp_after_insert_post` action, and will now have to do the same for all other actions.

So either we have to move those other actions, or use `meta_input` and address whatever the reasons were for not using that.

TimothyBJacobs replied:
The meta_input parameter is poorly designed for doing anything of complexity. It can't handle multi meta and support can't easily be added in a BC manner. It doesn't handle deleting meta. It doesn't provide any error handling. 

Perhaps the REST API could've been architected differently, but that ship has long since sailed. 

The REST API didn't move that hook, we introduced that hook specifically for the REST API. Core should be moving appropriate actions to that hook.

ellatrix replied:
It's not because something has been poorly designed that it can't be improved? Is there something fundamentally unfixable about `meta_input`?

Core could move actions to that hook, but that ship has also sailed? We'd need to make sure plugins removing these hooks are still removing the moved hooks somehow.

adamsilverstein replied:
@ellatrix I also agree this solution isn't fantastic, it is a bit of a compromise approach. The one part I do like is hooking `wp_save_revisioned_meta_fields` on an action which lets any plugin easily disable the feature entirely.

I don't quite understand the part about `meta_input`.  Are you saying the REST endpoint could populate that before updating the post? What would be a more ideal solution in your mind?  

I am open to considering changing this - of course we would need to be careful not to break any cases where plugins already expect the current behavior. Also, this change could easily be done **after** this PR lands, so I don't feel it should be a blocker here. We would still hook the revisioning as is, we would just remove the special unhooking/re-hooking logic in the posts controller if the order was fixed.

ellatrix replied:
> I don't quite understand the part about `meta_input`. Are you saying the REST endpoint could populate that before updating the post? What would be a more ideal solution in your mind?

What I mean is that there's a `meta` property that can be passed to the post data on the `posts` endpoint, which is essentially the same as `meta_input` that can be passed to the post data for `wp_insert_post` and `wp_update_post`. So there's great potential for symmetry here. I don't get the argument of it being poorly designed for complex things. The input for both should be almost the same: a keys to values array. So you're telling me that whatever handling the REST API is doing for meta, cannot be done inside of `wp_update_post`? Why can't we add deletion and error handling? If we do that, the action order will be correct and all will be well?

adamsilverstein replied:
> So you're telling me that whatever handling the REST API is doing for meta, cannot be done inside of wp_update_post? Why can't we add deletion and error handling? If we do that, the action order will be correct and all will be well?

Hmmm good point! I will test out making the change you are suggesting, I'm not sure why that wouldn't be possible maybe I am missing something and @TimothyBJacobs can clarify. I do see that there are [other changes](https://github.com/WordPress/wordpress-develop/blob/2acd0aca7845feda1c4ca4dd708ca4e8a72c422b/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L899-L918C26) also completed after `wp_update_post`, but I don't see why we can't just include the meta in `wp_update_post` as `meta_input` and - as you say - _all will be well_.

adamsilverstein committed:
add meta to revisions endpoint

adamsilverstein committed:
test revisions endpoint returns meta

adamsilverstein committed:
new revision meta fields class

adamsilverstein committed:
INCLUDE class-wp-rest-revision-meta-fields

adamsilverstein committed:
Revision existing meta even when update/insert lacks meta

adamsilverstein committed:
Test that the revisions endpoint includes meta

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

# Conflicts:
#	src/wp-includes/blocks/footnotes.php

adamsilverstein committed:
phpcbf

adamsilverstein commented:
> @adamsilverstein, should we try using `register_meta` for enabling post meta revisions instead of `wp_post_revision_meta_keys` filter?

@Mamaduka have it working way in the latest version (also filterable)

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
Update REST tests to reflect changed schema

adamsilverstein committed:
phpcbf

adamsilverstein committed:
Improve tests

adamsilverstein committed:
Finetune tests based on git action failures

adamsilverstein committed:
test

adamsilverstein committed:
restore footnotes.php from trunk

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
Update wp-api-generated fixtures

adamsilverstein committed:
fix generated

adamsilverstein committed:
Try: move hooking of revisions

Mamaduka commented on src/wp-includes/revision.php:
```diff
@@ -463,6 +487,85 @@ function wp_restore_post_revision( $revision, $fields = null ) {
+ */
+function _wp_copy_post_meta( $source_post_id, $target_post_id, $meta_key ) {
+
+	foreach ( get_post_meta( $source_post_id, $meta_key ) as $meta_value ) {
```
Can the `get_post_meta` return a non-array value? Should we safeguard it?

adamsilverstein replied:
good idea!

TimothyBJacobs replied:
Looking at the `get_metadata_raw` source, it's safe for us to assume this is an array.

adamsilverstein committed:
fire after hooks in revisions controller test

TimothyBJacobs committed:
Try only adding revision on update.

TimothyBJacobs committed:
Use actions properly

TimothyBJacobs committed:
Revert "Try only adding revision on update."

This reverts commit 96301dabc8de405528a028a076f0915bfa52f025.

TimothyBJacobs committed:
Don't rewrite everything

TimothyBJacobs committed:
Correct logic

TimothyBJacobs committed:
Check if the legacy hook is being used

TimothyBJacobs committed:
Separate single and multi meta tests. Use native meta fields handler.

TimothyBJacobs committed:
Regenerate fixtures

TimothyBJacobs committed:
Use post meta fields directly

TimothyBJacobs committed:
Respect the object subtype.

TimothyBJacobs committed:
Always pass a post array

TimothyBJacobs committed:
Remove manually revisioning meta. This happens automatically now.

TimothyBJacobs committed:
Correct has meta key assertion

TimothyBJacobs committed:
Add clarifying comment and some more fail safes

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein commented on src/wp-includes/revision.php:
```diff
@@ -132,7 +132,8 @@ function wp_save_post_revision( $post_id ) {
 
-	if ( doing_action( 'post_updated' ) ) {
+	// Prevent saving post revisions if revisions should be saved on wp_after_insert_post.
+	if ( doing_action( 'post_updated' ) && has_action( 'wp_after_insert_post', 'wp_save_post_revision_on_insert' ) ) {
```
Nice!

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

ellatrix commented:
Do we still have GB e2e tests in core? If they pass with this code, it's an instant approval for me.

adamsilverstein committed:
Test cleanup, add page and cpt tests

adamsilverstein commented:
> Do we still have GB e2e tests in core? If they pass with this code, it's an instant approval for me.

I'm not sure and will investigate. I should also be able to test locally by running the GB tests against core built with this patch.

adamsilverstein committed:
omyac

adamsilverstein commented:
I added additional tests for pages and a custom post type, final steps would be manual testing as well as getting Gutenberg E2E testing run to test this out.

Mamaduka commented:
I manually ran the Footnote block e2e test using the patch from https://github.com/WordPress/gutenberg/pull/52988 and this branch. Two tests related to revisions are failing.

## Screenshot
![CleanShot 2023-09-01 at 00 47 54](https://github.com/WordPress/wordpress-develop/assets/240569/3bab7fea-2d7e-4010-9680-23ed267a94a7)

ellatrix commented:
Thanks for running the tests!

adamsilverstein commented:
> I manually ran the Footnote block e2e test using the patch from WordPress/gutenberg#52988 and this branch. Two tests related to revisions are failing.

Excellent, thank you! 

I will try to reproduce the test steps manually to see what is happening. i'll also manually test all the features.

can you provide or point to instructions for how you ran the tests this way?  i can build core and GB, but i'm not sure how to change the container GB uses for test runs?

Mamaduka commented:
> can you provide or point to instructions for how you ran the tests this way

1. Clone the Gutenberg as a plugin in WP `trunk` installation.
2. Switch to the related branch in Gutenberg - https://github.com/WordPress/gutenberg/pull/52988 and build.
3. Switch to the related branch in the trunk and build.
4. Make sure that the WP trunk installation has a sample admin user (`admin/password`).
5. Run the following command in the Gutenberg plugin directory - `WP_BASE_URL=http://trunk.test/ npm run test:e2e:playwright -- test/e2e/specs/editor/various/footnotes.spec.js`
6. The `http://trunk.test/` should point to your local install. It can be `localhost` as well.

adamsilverstein commented:
Thanks! I'll get these running and fixed.

ellatrix commented:
Worth nothing Beta 1 is on 26 September, and the change in GB need to be done in the release before that. I'm just hoping this makes it in 6.4! :)

adamsilverstein commented:
> Worth nothing Beta 1 is on 26 September, and the change in GB need to be done in the release before that. I'm just hoping this makes it in 6.4! :)

Working on this again now!

@Mamaduka I'll work on testing manually; when I try running the tests in GB I'm stuck with an error:

![image](https://github.com/WordPress/wordpress-develop/assets/2676022/59528acb-8c76-4932-8d4b-b2a75c9d705e)

Any tips to get past that?

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

# Conflicts:
#	src/wp-includes/default-filters.php

adamsilverstein committed:
Clean up default filters after trunk merge

Mamaduka commented:
@adamsilverstein, you'll need to copy [`disable-animations.php`](https://raw.githubusercontent.com/WordPress/gutenberg/trunk/packages/e2e-tests/plugins/disable-animations.php) plugin.

adamsilverstein commented:
@Mamaduka thanks that fixed it for me and I was able to get the tests running. In headed mode I observed that the preview menu wasn't closing properly causing the next click to fail. I added a context click back to the first paragraph in [WordPress/gutenberg@`59bc54f` (#52988)](https://github.com/WordPress/gutenberg/pull/52988/commits/59bc54fe7c2ae347e55a5589fb9c3b82af1a27cb) and the tests passed for me after that.

Can you double check to verify tests all pass on your end?

I'm going to do some manual "smoke testing" to validate revisioning works correctly, then I think this is ready for commit!

adamsilverstein commented:
I did some manual testing as well, similar to the e2e test that was failing:

https://github.com/WordPress/wordpress-develop/assets/2676022/24134ae9-c245-43d7-b99c-83b07092125f


and a more extended session where I also restored a previous revision


https://github.com/WordPress/wordpress-develop/assets/2676022/e78112e1-be29-4571-a7ab-618d5832c801




everything worked as expected.

adamsilverstein committed:
Avoid duplicate revisions since hook change

Mamaduka commented:
Thank you, @adamsilverstein!

The `can be previewed when published` test is still failing for me, and I can reproduce the issue when testing manually.

I am unsure why this is happening, but the first autosave (generated when clicking the preview button) doesn't save changes to the footnotes meta. After I click preview a second time, all consecutive changes are correctly saved.

adamsilverstein commented:
Hmm. Was passing for me although I did make one additional change. I will give that test another try; thanks for testing

adamsilverstein commented:
Hmm, all tests pass for me - maybe something different in our environments? I'll try running all the tests.

![image](https://github.com/WordPress/wordpress-develop/assets/2676022/d8ffd098-3868-4804-bc95-1069bcc14e90)


> The can be previewed when published test is still failing for me, and I can reproduce the issue when testing manually.

> I am unsure why this is happening, but the first autosave (generated when clicking the preview button) doesn't save changes to the footnotes meta. After I click preview a second time, all consecutive changes are correctly saved.

I will try manually running through the steps to see if I can reproduce. Are you looking in the database or at the initial revision on the revisions screen - could it be an autosave?

Mamaduka commented:
Odd. It could be something on my local installation.

Can you double-check by following the steps from the test manually?

adamsilverstein commented:
> Can you double-check by following the steps from the test manually?

I did and it seemed to work fine.

Along the way, however, I discovered a couple of issues:
* The footnotes file here had been updated with some additional hooks since my last updated; I have a new commit to wrap those in conditionals as they are no longer needed
* I noticed a bug that is described in https://core.trac.wordpress.org/ticket/20299 - namely that post meta values on the live post are changed when previewing meta changes! I'm going to check to see if this is also true in WP 6.3, I suspect it may be.

adamsilverstein commented:
I verified the published post previewing meta bug already exists - still it would be good to fix this and add some tests top cover!

Testing on a vanilla 6.3.1 site with the latest GB plugin installed, I verified that previewing changes to post meta **also** writes that meta to the published post.

Here is a screencast, note that I publish the post, change the meta, then click preview - I don't ever update the post; nonetheless, the live post meta (footnotes) are updated:


https://github.com/WordPress/wordpress-develop/assets/2676022/ce605e17-597f-4b8d-8538-704c0e062fc8

adamsilverstein commented:
Testing in core with the `footnote.php` code updated with the latest from https://github.com/WordPress/gutenberg/pull/52988 resolves the overwriting of the live post meta, so I think this is a separate but that is introduced by one of the filters there.

adamsilverstein commented:
> Odd. It could be something on my local installation.

when manually testing, you need to copy the changes over from Gutenberg footnote block to core footnote.php - activating the plugin isn't enough to remove the current GB behavior.

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
Restore hooks on wp_insert_post for revision - issue was plugin related

adamsilverstein committed:
Ensure meta data is revisioned with autosaves

adamsilverstein committed:
Cleanup, only check meta when passed

Mamaduka commented:
I re-run the tests using the following method:

1. I deactivated Gutenberg on my trunk installation.
2. Copied PHP changes from https://github.com/WordPress/gutenberg/pull/52988 into `src/wp-includes/blocks/footnotes.php`.
3. Run unit tests using - `WP_BASE_URL=http://trunk.test/ npm run test:e2e:playwright -- test/e2e/specs/editor/various/footnotes.spec.js`

## Result

All tests pass except the last assertion in `can be previewed when published`:

```js
// Note: quote will get curled by wptexturize.
await expect(
	previewPage2.locator( 'ol.wp-block-footnotes li' )
).toHaveText( '123″  ↩︎' );
```

Here's the [trace archvie](https://github.com/WordPress/wordpress-develop/files/12595745/trace.zip) from my local run; it can be viewed via https://trace.playwright.dev/.

> activating the plugin isn't enough to remove the current GB behavior.

We'll need to temporarily remove old callbacks in the plugin before changes from https://github.com/WordPress/gutenberg/pull/52988 are synced with the core.

adamsilverstein commented:
> I re-run the tests using the following method:
 
Thanks for testing. 

Odd that test fails for you, it works for me! 

I wasn't able to open the trace. If you try the steps manually, what do you see on the preview page, is the meta not updated?

// Copied PHP changes from Footnotes: use core’s meta revisioning if available gutenberg#52988 into src/wp-includes/blocks/footnotes.php.

I assume you did an `grunt build` after this or are running from src?

I will plan to double check functionality manually, then commit this before beta so it can get wider testing. Once it is in core, we can make sure all the GB tests pass.

I tried working on adding similar tests in core directly, but was frustrated by core's use of Puppeteer.

Mamaduka commented:
I'm still having a problem successfully running the `can be previewed when published` e2e test locally. However, everything is working correctly when testing manually, so it's probably an issue with my local setup.

Sorry for the false alarm, @adamsilverstein!

> I will plan to double check functionality manually, then commit this before beta so it can get wider testing. Once it is in core, we can make sure all the GB tests pass.

Sounds good to me ✅

> I tried working on adding similar tests in core directly, but was frustrated by core's use of Puppeteer.

We need to switch core to use Playwright as well; then we can just copy-paste tests as needed.

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein commented:
> I'm still having a problem successfully running the `can be previewed when published` e2e test locally. However, everything is working correctly when testing manually, so it's probably an issue with my local setup.
 
Excellent, good to hear! Thanks for the manual testing.

> > I tried working on adding similar tests in core directly, but was frustrated by core's use of Puppeteer.
> 
> We need to switch core to use Playwright as well; then we can just copy-paste tests as needed.

That would be fantastic, especially for cases like this where a core change needs to sync with a Gutenberg change!

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

Mamaduka commented:
@adamsilverstein, are you planning to commit this before beta 1 in this release cycle?

spacedmonkey commented on src/wp-admin/includes/post.php:
```diff
@@ -1961,15 +1961,75 @@ function wp_create_post_autosave( $post_data ) {
 		 * @param array $new_autosave Post array - the autosave that is about to be saved.
 		 */
 		do_action( 'wp_creating_autosave', $new_autosave );
-
+		wp_autosave_post_revisioned_meta_fields( $new_autosave );
```
Wonder if we could hook `wp_autosave_post_revisioned_meta_fields` into like this. 

```
add_action( 'wp_creating_autosave', 'wp_autosave_post_revisioned_meta_fields' );
```

To make this unhookable...

adamsilverstein replied:
Great idea!

adamsilverstein replied:
Added in [`750dcdc` (#4859)](https://github.com/WordPress/wordpress-develop/pull/4859/commits/750dcdc81cbcc1de7085d776516e87fef9c70938)

spacedmonkey commented on src/wp-admin/includes/post.php:
```diff
@@ -1961,15 +1961,75 @@ function wp_create_post_autosave( $post_data ) {
 
 	// Otherwise create the new autosave as a special post revision.
-	return _wp_put_post_revision( $post_data, true );
+	$revision = _wp_put_post_revision( $post_data, true );
```
This function can return a WP_Error. We should have some error handling here.

adamsilverstein replied:
👍🏼 added

spacedmonkey commented on src/wp-admin/includes/post.php:
```diff
@@ -1961,15 +1961,75 @@ function wp_create_post_autosave( $post_data ) {
+ *
+ * @param array $new_autosave The new post data being autosaved.
+ */
+function wp_autosave_post_revisioned_meta_fields( $new_autosave ) {
```
Why does this need to be a array and not a WP_Post object?

adamsilverstein replied:
The `wp_creating_autosave` hook where this is now hooked passes a post array. previously I was calling with that array and would have needed to transform into a post object.

adamsilverstein replied:
Is there some reason it would be preferable as a Post object?

spacedmonkey replied:
Just interested.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -390,11 +406,21 @@ public function create_post_autosave( $post_data ) {
 
 			// wp_update_post() expects escaped array.
-			return wp_update_post( wp_slash( $new_autosave ) );
+			$revision_id = wp_update_post( wp_slash( $new_autosave ) );
+		} else {
+			// Create the new autosave as a special post revision.
+			$revision_id = _wp_put_post_revision( $post_data, true );
```
Both `wp_update_post` and `_wp_put_post_revision` can return a WP_Error object. 

There NEEDS to be error handling here.

adamsilverstein replied:
I was checking `&& $revision_id` in the loop. I added explicit error checking with `if ( ! empty( $meta ) && $revision_id && ! is_wp_error( $revision_id ) ) {`. The loop will be skipped on an error condition and the error returned to be handled by the caller which is the expected behavior.

spacedmonkey commented on src/wp-includes/revision.php:
```diff
@@ -361,15 +387,36 @@ function _wp_put_post_revision( $post = null, $autosave = false ) {
+function wp_save_revisioned_meta_fields( $revision_id, $post_id ) {
+
+	// Save revisioned meta fields.
+	foreach ( wp_post_revision_meta_keys( get_post_type( $post_id ) ) as $meta_key ) {
```
get_post_type( $post_id ) could return false. Error handling?

adamsilverstein replied:
I'm not sure about this one. Under what conditions would the post_type be false? Would we want to skip the check entirely?

I assume we would exit early on an error here, however do we need to? Passing false to the `wp_post_revision_meta_keys` function will still work, it will use meta types from 'post' and pass the `false` type to the filter.

hellofromtonya replied:
It would be `false` if `get_post()` returns a falsey value. That would typically happen the post ID is falsey or that ID doesn't exist.

In the normal run, the new function is a hooked callback and the post ID is from the post's array. That said, is there a possible scenario where the `'post_parent'` post does not exist in the database?

Also what happens if this function is directly called by an extender? Safeguarding would protect against a falsey state.

hellofromtonya replied:
```
foreach ( wp_post_revision_meta_keys( get_post_type( $post_id ) )
```

I'd also suggest breaking out the compound code from the `foreach()` to make the code more understandable for all contributors.

```php
$post_type = get_post_type( $post_id );
if ( ! $post_type ) {
   return;
}

foreach ( wp_post_revision_meta_keys( $post_type ) as $meta_key )  {
```

adamsilverstein replied:
Makes sense, thanks for the feedback.

adamsilverstein replied:
Note, also updating in `wp_restore_post_revision_meta()`

adamsilverstein replied:
updated

spacedmonkey commented on src/wp-includes/revision.php:
```diff
@@ -450,6 +497,9 @@ function wp_restore_post_revision( $revision, $fields = null ) {
 	update_post_meta( $post_id, '_edit_last', get_current_user_id() );
 
+	// Restore any revisioned meta fields.
+	wp_restore_post_revision_meta( $post_id, $revision['ID'] );
```
Could we do this 

```
add_action( 'wp_restore_post_revision', 'wp_restore_post_revision_meta', 10, 2 );
```

adamsilverstein replied:
good idea, let me check.

adamsilverstein replied:
Great! Added in [`95508d9` (#4859)](https://github.com/WordPress/wordpress-develop/pull/4859/commits/95508d959d69e6be174c922c8308dac51077b7ff)

spacedmonkey replied:
@adamsilverstein Has this been addressed?

adamsilverstein replied:
yes, this is added

spacedmonkey commented on src/wp-includes/revision.php:
```diff
@@ -463,6 +513,101 @@ function wp_restore_post_revision( $revision, $fields = null ) {
+function wp_restore_post_revision_meta( $post_id, $revision_id ) {
+
+	// Restore revisioned meta fields.
+	foreach ( wp_post_revision_meta_keys( get_post_type( $post_id ) ) as $meta_key ) {
```
https://github.com/WordPress/wordpress-develop/pull/4859/files#r1335858255

adamsilverstein replied:
see above, maybe fine to pass false here?

spacedmonkey commented on src/wp-includes/revision.php:
```diff
@@ -463,6 +513,101 @@ function wp_restore_post_revision( $revision, $fields = null ) {
+ * @since 6.4.0
+ */
+function wp_check_revisioned_meta_fields_have_changed( $post_has_changed, WP_Post $last_revision, WP_Post $post ) {
+	foreach ( wp_post_revision_meta_keys( $post->post_type ) as $meta_key ) {
```
What happens here if the is a default value set on either of these?

adamsilverstein replied:
I will check (and add a unit test to cover)

adamsilverstein replied:
Added tests: [`b0b880b` (#4859)](https://github.com/WordPress/wordpress-develop/pull/4859/commits/b0b880be91d935c27353631537d439a1406e7d35)

spacedmonkey commented on src/wp-includes/revision.php:
```diff
@@ -946,3 +1092,38 @@ function _wp_upgrade_revisions_of_post( $post, $revisions ) {
+	}
+
+	$preview = wp_get_post_autosave( $post->ID );
+	if ( ! is_object( $preview ) ) {
```
```suggestion
	if ( false === $preview ) {
```

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -24,6 +24,14 @@ class WP_REST_Revisions_Controller extends WP_REST_Controller {
+	/**
+	 * Instance of a revision meta fields object.
+	 *
+	 * @since 4.7.0
```
```suggestion
	 * @since 6.4.0
```

spacedmonkey commented on src/wp-includes/meta.php:
```diff
@@ -1414,6 +1417,7 @@ function register_meta( $object_type, $meta_key, $args, $deprecated = null ) {
 		'sanitize_callback' => null,
 		'auth_callback'     => null,
 		'show_in_rest'      => false,
+		'revisions_enabled' => false,
```
We could add a `_doing_it_wrong` if you pass `revisions_enabled` and type that was not post. We could even add a message saying that post type does not support revisions.

adamsilverstein replied:
I think you mean for a post type that doesn't support revisions which any post type can opt into with supports=>revisions. makes sense.

adamsilverstein replied:
Added a check for this condition in [`f9b760b` (#4859)](https://github.com/WordPress/wordpress-develop/pull/4859/commits/f9b760b7d5fb9de3e56e4ebadb60c9fcda4c6ccd)

Added a test to validate the check works correctly in [`8360faa` (#4859)](https://github.com/WordPress/wordpress-develop/pull/4859/commits/8360faa26e9aa07718aba603d37f62f70039615f)

adamsilverstein replied:
Now emits `Meta keys cannot enable revisions support unless the object type supports revisions.`

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -372,6 +374,20 @@ public function create_post_autosave( $post_data ) {
+		if ( ! empty( $meta ) ) {
+			$revisioned_meta_keys = wp_post_revision_meta_keys( $post->post_type );
+			foreach ( $revisioned_meta_keys as $meta_key ) {
+				$old_meta = get_post_meta( $post_id, $meta_key, true );
```
What if there is a default value is set for this key? Should we be using `get_metadata_raw` here?

adamsilverstein replied:
good question, I'm not sure what the behavior should be for meta with default values. I will think on this further, and we should certainly add tests here to verify it behaves as we expect.

adamsilverstein replied:
Good catch, changed.

adamsilverstein replied:
Going to check this manually before pushing since our unit tests don't really cover this.

adamsilverstein replied:
Updated this in [`e1950b0` (#4859)](https://github.com/WordPress/wordpress-develop/pull/4859/commits/e1950b0debef4eeb629f932a9dbe0dbe81cd6a16)

spacedmonkey reviewed (changes requested):
This needs better error handling. There are places where functions can return WP_error that are not handled.

adamsilverstein commented:
> This needs better error handling. There are places where functions can return WP_error that are not handled.

Good points and thanks for the review! will address before commit.

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
Hook wp_autosave_post_revisioned_meta_fields on wp_creating_autosave

Also, add a second triggering of `wp_creating_autosave` and add a parameter indicating if this is an update or a new autosave revision.

adamsilverstein committed:
Improve error checking

adamsilverstein committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

adamsilverstein committed:
Update src/wp-includes/revision.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

adamsilverstein committed:
Hook revision restore on action

adamsilverstein commented on src/wp-includes/revision.php:
```diff
@@ -463,6 +513,101 @@ function wp_restore_post_revision( $revision, $fields = null ) {
+ */
+function _wp_copy_post_meta( $source_post_id, $target_post_id, $meta_key ) {
+
+	foreach ( get_post_meta( $source_post_id, $meta_key ) as $meta_value ) {
```
@spacedmonkey - I think the same issue for defaults applies here. If a default is set for a meta field, this code will copy the _default_ value over instead of the actual (missing) value, which is unintentional. switching is to get_metadata_raw as well (although may have been prefented by check for existence of meta in calling function)

adamsilverstein replied:
Changing this to `get_metadata_raw( 'post', $source_post_id, $meta_key = '' )` broke numerous revision group tests so leaving as is. Could use more investigation, in any case this function is not called when the meta value is not set.

adamsilverstein committed:
Test revisioning of meta with a default value

adamsilverstein committed:
phpcs

adamsilverstein committed:
phpcbf

hellofromtonya commented on src/wp-includes/revision.php:
```diff
@@ -361,15 +387,36 @@ function _wp_put_post_revision( $post = null, $autosave = false ) {
+ * @param int $post_id     The ID of the post the revision is associated with.
+ *
+ * @since 6.4.0
+ */
```
```suggestion
/**
 * Save the revisioned meta fields.
 *
 * @since 6.4.0
 *
 * @param int $revision_id The ID of the revision to save the meta to.
 * @param int $post_id     The ID of the post the revision is associated with.
 */
```

adamsilverstein committed:
Autosaves controller - get_metadata_raw is used to avoid retrieving the default value

adamsilverstein committed:
Update src/wp-includes/revision.php

Correct doc block order

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

adamsilverstein committed:
Check post_type available in wp_save_revisioned_meta_fields

adamsilverstein committed:
Ensure post_type available in wp_restore_post_revision_meta

adamsilverstein commented:
> @adamsilverstein, are you planning to commit this before beta 1 in this release cycle?

Yes

adamsilverstein committed:
typo - missing closing parenthesis

adamsilverstein committed:
phpcbf

adamsilverstein committed:
If the object_type is set, require a type that supports revisions when setting revisions_enabled to true

adamsilverstein committed:
Test enabling revisions for meta when object type doesn’t support revisions

adamsilverstein committed:
phpcbf

spacedmonkey commented on src/wp-includes/meta.php:
```diff
@@ -1459,6 +1463,13 @@ function register_meta( $object_type, $meta_key, $args, $deprecated = null ) {
 		}
 	}
 
+	// If the object_type is set, require a type that supports revisions when setting revisions_enabled to true.
+	if ( ! empty( $object_type ) && $args['revisions_enabled'] && ! post_type_supports( $object_type, 'revisions' ) ) {
+		_doing_it_wrong( __FUNCTION__, __( 'Meta keys cannot enable revisions support unless the object type supports revisions.' ), '6.4.0' );
+
+		return false;
+	}
+
 	$object_subtype = ! empty( $args['object_subtype'] ) ? $args['object_subtype'] : '';
```
```suggestion
	$object_subtype = ! empty( $args['object_subtype'] ) ? $args['object_subtype'] : '';
	if( $args['revisions_enabled'] ) {
	     if( 'post' !== $object_type ) {
	       _doing_it_wrong( __FUNCTION__, __( 'Meta keys cannot enable revisions support unless the object type supports revisions.' ), '6.4.0' );

		  return false;
	     } else if ( ! empty( $object_subtype ) && ! post_type_supports( $object_subtype, 'revisions' ) ) {
		_doing_it_wrong( __FUNCTION__, __( 'Meta keys cannot enable revisions support unless the object type supports revisions.' ), '6.4.0' );

		return false;
	    }
	}
```

adamsilverstein replied:
I'm not sure this works. If the object type is empty (""), the meta is for ALL post types, but your logic will throw doing_it_wrong in this case.

spacedmonkey replied:
@adamsilverstein To be clear, there is object type and $object_subtype. Object type can be user, post, term etc. Sub type is for post types, like page. 

The two checks are designed to do this. 

- If user try assign `revisions_enabled` enabled on say users, return a doing it wrong. 
- If type of post is set and object subtype is set and subtype doesnt support revisions, return a doing it wrong.

adamsilverstein replied:
Oh, thanks for clarifying! I totally missed that you were checking the subtype when reviewing the suggestion here on GitHub.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -348,11 +348,13 @@ public function get_item_schema() {
 	 * @return mixed The autosave revision ID or WP_Error.
 	 */
-	public function create_post_autosave( $post_data ) {
+	public function create_post_autosave( $post_data, $meta = array() ) {
```
```suggestion
	public function create_post_autosave( $post_data, array $meta = array() ) {
```

We can do this, which is neat.

adamsilverstein replied:
Neat, does this enforce passing an array?

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -390,11 +407,21 @@ public function create_post_autosave( $post_data ) {
 
-		// Create the new autosave as a special post revision.
-		return _wp_put_post_revision( $post_data, true );
+		// Attached any passed meta values that have revisions enabled.
+		if ( ! empty( $meta ) && $revision_id && ! is_wp_error( $revision_id ) ) {
+			foreach ( $revisioned_meta_keys as $meta_key ) {
+				if ( isset( $meta[ $meta_key ] ) ) {
+					update_metadata( 'post', $revision_id, $meta_key, $meta[ $meta_key ] );
+				}
+			}
+		}
+		return $revision_id;
```
```suggestion
                 if( is_wp_error( $revision_id ) ) {
                      return $revision_id;
                 }
                 
		// Attached any passed meta values that have revisions enabled.
		if ( ! empty( $meta ) && $revision_id ) {
			foreach ( $revisioned_meta_keys as $meta_key ) {
				if ( isset( $meta[ $meta_key ] ) ) {
					update_metadata( 'post', $revision_id, $meta_key, $meta[ $meta_key ] );
				}
			}
		}
		
		return $revision_id;
```

adamsilverstein replied:
I like the return early, nice. Looks like you have spaces instead of tabs here. I'mm merge this and fix in a follow up.

spacedmonkey commented:
Pinging @swissspidy as this change may effect the web stories plugin that has a custom revision / autosave endpoint.

spacedmonkey reviewed (changes requested):
I tested this with the web stories plugin that has many custom post types and meta. Seems to work well. Just some little bits of feedback and we should be nearly ready to commit.

swissspidy commented on src/wp-admin/includes/post.php:
```diff
@@ -1957,19 +1957,88 @@ function wp_create_post_autosave( $post_data ) {
+
+	if ( ! is_wp_error( $revision ) ) {
+
+		/**
+		 * Fires before an autosave is stored.
+		 *
+		 * This filter is documented in `src/wp-admin/includes/post.php`.
+		 */
```
```suggestion
		/** This action is documented in wp-admin/includes/post.php */
```

swissspidy commented on src/wp-admin/includes/post.php:
```diff
@@ -1957,19 +1957,88 @@ function wp_create_post_autosave( $post_data ) {
+/**
+ * Autosave the revisioned meta fields.
+ *
+ * Iterates thru the revisioned meta fields and checks each to see if they are set,
```
```suggestion
 * Iterates through the revisioned meta fields and checks each to see if they are set,
```

adamsilverstein committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

adamsilverstein committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

adamsilverstein committed:
Update src/wp-admin/includes/post.php

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

adamsilverstein committed:
Update src/wp-admin/includes/post.php

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

adamsilverstein committed:
spacing

adamsilverstein commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -390,11 +407,26 @@ public function create_post_autosave( $post_data ) {
+			$revision_id = _wp_put_post_revision( $post_data, true );
+		}
+
+		if( is_wp_error( $revision_id ) ) {
```
@spacedmonkey thinking we should check for 0 here as well since `_wp_put_post_revision` can return 0.

spacedmonkey replied:
How about this?
```
if( 0 === $revision_id ) {
    return new WP_Error(...
}
```

adamsilverstein replied:
$revision_id could be 0 OR a wp_error here so I think we need to check for both which is what I did.

adamsilverstein committed:
Guard against _wp_put_post_revision returning 0

adamsilverstein committed:
Update src/wp-includes/meta.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

adamsilverstein committed:
tabs vs spaces

spacedmonkey commented on src/wp-includes/meta.php:
```diff
@@ -1460,6 +1464,17 @@ function register_meta( $object_type, $meta_key, $args, $deprecated = null ) {
+
+		  return false;
+	     } else if ( ! empty( $object_subtype ) && ! post_type_supports( $object_subtype, 'revisions' ) ) {
+		_doing_it_wrong( __FUNCTION__, __( 'Meta keys cannot enable revisions support unless the object type supports revisions.' ), '6.4.0' );
```
```suggestion
		_doing_it_wrong( __FUNCTION__, __( 'Meta keys cannot enable revisions support unless the object subtype supports revisions.' ), '6.4.0' );
```

adamsilverstein replied:
thanks, I'll update this in my local as github won't let me merge here

spacedmonkey commented:
Do we need to update `wp_create_initial_post_meta`?

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -234,8 +234,8 @@ public function create_item( $request ) {
-			// Non-draft posts: create or update the post autosave.
-			$autosave_id = $this->create_post_autosave( (array) $prepared_post );
+			// Non-draft posts: create or update the post autosave. Pass the meta data.
+			$autosave_id = $this->create_post_autosave( (array) $prepared_post, $request->get_param( 'meta' ) );
```
```suggestion
			$autosave_id = $this->create_post_autosave( (array) $prepared_post, (array) $request->get_param( 'meta' ) );
```

@adamsilverstein that should fix the unit tests.

adamsilverstein committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

adamsilverstein committed:
Improve error message

adamsilverstein commented:
> Do we need to update wp_create_initial_post_meta?

Not sure, I'm going to get this committed then we can revisit during beta.

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
Merge branch 'trunk' into ticket/20564

adamsilverstein committed:
phpcbf

adamsilverstein committed:
phpcbf pt 2

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -390,11 +407,26 @@ public function create_post_autosave( $post_data ) {
+		}
+
+		// Attached any passed meta values that have revisions enabled.
+		if ( ! empty( $meta ) && $revision_id ) {
```
This will always be not empty right? `$revision_id `

adamsilverstein replied:
Right, this will always be set now, I can remove

adamsilverstein replied:
Updated

spacedmonkey commented on src/wp-includes/revision.php:
```diff
@@ -450,6 +500,9 @@ function wp_restore_post_revision( $revision, $fields = null ) {
 	// Update last edit user.
 	update_post_meta( $post_id, '_edit_last', get_current_user_id() );
 
+	// Restore any revisioned meta fields.
+	wp_restore_post_revision_meta( $post_id, $revision['ID'] );
```
```suggestion
```

@adamsilverstein This is not need anymore, as it hooked in the next line.

adamsilverstein replied:
👍🏼 will commit in a follow up

spacedmonkey reviewed (approved):
Good to commit. We will follow up with any issues.

adamsilverstein commented:
Fixed in https://core.trac.wordpress.org/changeset/56714
